### PR TITLE
Add PathBuf return to from_filename() and dotenv()

### PR DESCRIPTION
**Pull request by daboross**
This implements #22, with the only differences being returning PathBufs instead of Strings, and not adding the return type to from_path.

My reasoning for choosing PathBuf over String is that we already create a PathBuf when using `.join()` in try_parent and from_filename, thus returning that PathBuf will be cheaper than allocating a new String to return. This way there's no (even small) disadvantage for users who do not use this feature!

Hope this is helpful, any feedback is appreciated.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -230,13 +230,14 @@ fn from_file(file: File) -> Result<(), DotenvError> {
...
     match path.parent() {
         Some(parent) => {
-            match from_path(&parent.join(filename)) {
-                Ok(file) => Ok(file),
+            let env_path = parent.join(filename);
+            match from_path(&env_path) {
+                Ok(()) => Ok(env_path),
                 Err(DotenvError::Io(_)) => try_parent(parent, filename),
-                err => err
+                Err(other) => Err(other)
```
Comment (daboross) [inline src/lib.rs]: Just nothing that this change is in order to turn a `Result<(), DotenvError>` into a `Result<PathBuf, DotenvError>` - same with the change in `from_filename`.

**Comment by slapresta** [comment]
@daboross Thank you so much for implementing this!

I had a try at solving the conflicts in the branch before merging. Could you quickly verify that the end result is correct? Thanks!

**Comment by daboross** [comment]
Glad to help!

I think this will need a few more changes for the new errors. I'll add that now - would you prefer a clean rebase or another commit on top of the merge?

**Comment by slapresta** [comment]
Nice! Merging 👍

**Comment by daboross** [comment]
Thanks!

**Pull request merged**